onnxruntime/core/providers/nnapi/nnapi_builtin/builders/impl/gather_op_builder.cc

```diff
@@ -98,12 +99,21 @@ Status GatherOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder, const
       indices.assign(indice_span.begin(), indice_span.end());
     }
 
+    // If indices is a scalar, insert a dimension of 1
+    // ONNX spec says that indices can be q-D tensor, q could be any value >= 0
+    // When indices is a scalar, output shape would be automatically squeezed.
+    // Add a fake output dimension to avoid NNAPI error.
+    if (indices_dimen.empty()) {
+      indices_dimen.push_back(1);
+      output_shape.insert(output_shape.begin() + axis, 1);
+    }
+
     OperandType indices_operand_type(Type::TENSOR_INT32, indices_dimen);
     ORT_RETURN_IF_ERROR(model_builder.AddOperandFromPersistMemoryBuffer(input2, indices.data(), indices_operand_type));
   }
   input_indices.push_back(operand_indices.at(input2));
 
-  const OperandType output_operand_type(operand_types.at(input1).type, shaper[output]);
+  const OperandType output_operand_type(operand_types.at(input1).type, output_shape);
```

Comment:
This needs unit tests to validate. Naively I would have expected a Squeeze needed to be added to remove the axis dim from the output, otherwise the output shape is different to what is happening in the ONNX model.

e.g. if axis is 1 and input shape was {3, 3, 4} the NNAPI output shape will now be {3, 1, 4 } but the ONNX Reshape output will be {3, 4} so we need to remove dim 1 from the NNAPI output to make it equivalent.

Reply:
Yeah, we have to match the output shape and next node's input shape.